File: src/main/java/io/jenkins/plugins/gamekins/GameJobPropertyDescriptor.java
package io.jenkins.plugins.gamekins;

import hudson.Extension;
import hudson.model.*;
import hudson.util.FormValidation;
import hudson.util.ListBoxModel;
import org.kohsuke.stapler.AncestorInPath;
import org.kohsuke.stapler.QueryParameter;

import javax.annotation.Nonnull;
import java.io.IOException;

@Extension
public class GameJobPropertyDescriptor extends JobPropertyDescriptor {

    public GameJobPropertyDescriptor() {
        super(GameJobProperty.class);
        load();
    }

    @Nonnull
    @Override
    public String getDisplayName() {
        return "Set the activation of the Gamekins plugin.";
    }

    @Override
    public boolean isApplicable(Class<? extends Job> jobType) {
        //TODO: Only for FreeStyle and WorkflowJob
        return true;
    }

    public FormValidation doAddTeam(@AncestorInPath Job<?, ?> job, @QueryParameter String teamName) {
        if (teamName.isEmpty()) return FormValidation.error("Insert a name for the team");
        GameJobProperty property = job == null ? null : (GameJobProperty) job.getProperties().get(this);
        if (property == null || property.getTeams() == null) return FormValidation.error("Unexpected Error");
        if (property.getTeams().contains(teamName))
            return FormValidation.error("The team already exists - please use another name for your team");
        try {
            property.addTeam(teamName);
        } catch (IOException e) {
            e.printStackTrace();
            return FormValidation.error("Unexpected Error");
        }
        save();
        return FormValidation.ok();
    }

    public ListBoxModel doFillTeamsBoxItems(@AncestorInPath Job<?, ?> job) {
        GameJobProperty property = job == null ? null : (GameJobProperty) job.getProperties().get(this);
        ListBoxModel listBoxModel = new ListBoxModel();
        if (property != null && property.getTeams() != null) property.getTeams().forEach(listBoxModel::add);
        return listBoxModel;
    }

    public ListBoxModel doFillUsersBoxItems() {
        ListBoxModel listBoxModel = new ListBoxModel();
        User.getAll().stream().map(User::getFullName).forEach(listBoxModel::add);
        listBoxModel.remove("unknown");
        return listBoxModel;
    }

    public FormValidation doAddUserToTeam(@AncestorInPath Job<?, ?> job, @QueryParameter String teamsBox, @QueryParameter String usersBox) {
        for (User user : User.getAll()) {
            if (user.getFullName().equals(usersBox)) {
                String projectName = job.getName();
                GameUserProperty property = user.getProperty(GameUserProperty.class);
                if (property != null && !property.isParticipating(projectName)) {
                    property.setParticipating(projectName, teamsBox);
                    try {
                        user.save();
                    } catch (IOException e) {
                        return FormValidation.error(e, "There was an error with saving");
                    }
                    return FormValidation.ok();
                } else {
                    return FormValidation.error("The user is already participating in a team");
                }
            }
        }
        return FormValidation.error("No user with the specified name found");
    }

    public FormValidation doRemoveUserFromTeam(@AncestorInPath Job<?, ?> job, @QueryParameter String teamsBox, @QueryParameter String usersBox) {
        for (User user : User.getAll()) {
            if (user.getFullName().equals(usersBox)) {
                String projectName = job.getName();
                GameUserProperty property = user.getProperty(GameUserProperty.class);
                if (property != null && property.isParticipating(projectName, teamsBox)) {
                    property.removeParticipation(projectName);
                    try {
                        user.save();
                    } catch (IOException e) {
                        return FormValidation.error(e, "There was an error with saving");
                    }
                    return FormValidation.ok();
                } else {
                    return FormValidation.error("The user is not in the specified team");
                }
            }
        }
        return FormValidation.error("No user with the specified name found");
    }

    public FormValidation doDeleteTeam(@AncestorInPath Job<?, ?> job, @QueryParameter String teamsBox) {
        String projectName = job.getName();
        GameJobProperty jobProperty = (GameJobProperty) job.getProperties().get(this);
        if (jobProperty == null || jobProperty.getTeams() == null) return FormValidation.error("Unexpected Error");
        if (!jobProperty.getTeams().contains(teamsBox)) return FormValidation.error("The specified team does not exist");
        for (User user : User.getAll()) {
            GameUserProperty property = user.getProperty(GameUserProperty.class);
            if (property != null && property.isParticipating(projectName, teamsBox)) {
                property.removeParticipation(projectName);
                try {
                    user.save();
                } catch (IOException e) {
                    return FormValidation.error(e, "There was an error with saving");
                }
            }
        }
        try {
            jobProperty.removeTeam(teamsBox);
        } catch (IOException e) {
            e.printStackTrace();
            return FormValidation.error("Unexpected Error");
        }
        save();
        return FormValidation.ok();
    }
}

File: src/main/kotlin/org/gamekins/property/GameJobPropertyDescriptor.kt
/*
 * Copyright 2023 Gamekins contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gamekins.property

import hudson.Extension
import hudson.maven.AbstractMavenProject
import hudson.model.*
import hudson.util.FormValidation
import hudson.util.ListBoxModel
import org.gamekins.util.PropertyUtil
import net.sf.json.JSONObject
import org.gamekins.util.Constants
import org.jenkinsci.plugins.workflow.job.WorkflowJob
import org.kohsuke.stapler.AncestorInPath
import org.kohsuke.stapler.QueryParameter
import org.kohsuke.stapler.StaplerRequest
import javax.annotation.Nonnull

/**
 * Registers the [GameJobProperty] to Jenkins as an extension and also works as an communication point between the
 * Jetty server and the [GameJobProperty].
 *
 * @author Philipp Straubinger
 * @since 0.1
 */
@Extension
class GameJobPropertyDescriptor : JobPropertyDescriptor(GameJobProperty::class.java) {

    /**
     * Called from the Jetty server if the button to add a new team is pressed. Only allows a non-empty [teamName] and
     * adds them to the [job], from which the button has been clicked, via the method [PropertyUtil.doAddTeam].
     */
    fun doAddTeam(@AncestorInPath job: Job<*, *>?, @QueryParameter teamName: String): FormValidation {
        if (teamName.isEmpty()) return FormValidation.error(Constants.Error.NO_TEAM_NAME)
        val property = if (job == null) null else job.properties[this] as GameJobProperty?
        val validation = PropertyUtil.doAddTeam(property, teamName)
        save()
        return validation
    }

    /**
     * Called from the Jetty server if the button to add a new participant to a team is pressed. Adds the participant
     * [usersBox] to the team [teamsBox] via the method [PropertyUtil.doAddUserToTeam].
     */
    fun doAddUserToTeam(@AncestorInPath job: Job<*, *>?, @QueryParameter teamsBox: String?,
                        @QueryParameter usersBox: String?): FormValidation {
        return PropertyUtil.doAddUserToTeam(job, teamsBox!!, usersBox!!)
    }

    /**
     * Called from the Jetty server if the button to delete a team is pressed. Deletes the team [teamsBox] of the
     * [job] via the method [PropertyUtil.doDeleteTeam].
     */
    fun doDeleteTeam(@AncestorInPath job: Job<*, *>?, @QueryParameter teamsBox: String?): FormValidation {
        if (job == null) return FormValidation.error(Constants.Error.PARENT)
        val projectName = job.fullName
        val property = job.properties[this] as GameJobProperty
        val validation = PropertyUtil.doDeleteTeam(projectName, property, teamsBox!!)
        save()
        return validation
    }

    /**
     * Called from the Jetty server when the configuration page is displayed. Fills the combo box with the names of
     * all teams of the [job].
     */
    fun doFillTeamsBoxItems(@AncestorInPath job: Job<*, *>?, @QueryParameter includeNoTeam: Boolean): ListBoxModel {
        val property =
                if (job == null || job.properties[this] == null) null
                else job.properties[this] as GameJobProperty?
        return PropertyUtil.doFillTeamsBoxItems(property, includeNoTeam)
    }

    /**
     * Called from the Jetty server when the configuration page is displayed. Fills the combo box with the names of
     * all users of the [job].
     */
    fun doFillUsersBoxItems(@AncestorInPath job: Job<*, *>?): ListBoxModel {
        return if (job == null) ListBoxModel() else PropertyUtil.doFillUsersBoxItems(job.fullName)
    }

    /**
     * Called from the Jetty server if the button to remove a participant from a team is pressed. Removes the
     * participant [usersBox] from the project of the [job] via the method [PropertyUtil.doRemoveUserFromProject].
     */
    fun doRemoveUserFromProject(@AncestorInPath job: Job<*, *>?, @QueryParameter usersBox: String?): FormValidation {
        return PropertyUtil.doRemoveUserFromProject(job, usersBox!!)
    }

    /**
     * Called from the Jetty server if the button to participate alone is pressed. Adds the participant
     * [usersBox] to a Pseudo-Team via the method [PropertyUtil.doAddUserToTeam].
     */
    fun doParticipateAlone(@AncestorInPath job: Job<*, *>?, @QueryParameter usersBox: String?): FormValidation {
        return PropertyUtil.doAddUserToTeam(job, Constants.NO_TEAM_TEAM_NAME, usersBox!!)
    }

    /**
     * Called from the Jetty server if the button to reset Gamekins in the specific [job] is pressed. Deletes all
     * Challenges from all users for the current project and resets the statistics.
     */
    fun doReset(@AncestorInPath job: Job<*, *>?): FormValidation {
        val property =
                if (job == null || job.properties[this] == null) null
                else job.properties[this] as GameJobProperty?
        return PropertyUtil.doReset(job, property)
    }

    /**
     * Called from the Jetty server if the button to show the team memberships in the specific [job] is pressed.
     * Returns a map of teams and their members as json.
     */
    fun doShowTeamMemberships(@AncestorInPath job: Job<*, *>?): String {
        val property =
            if (job == null || job.properties[this] == null) return ""
            else job.properties[this] as GameJobProperty
        return PropertyUtil.doShowTeamMemberships(job, property)
    }

    @Nonnull
    override fun getDisplayName(): String {
        return "Set the activation of the Gamekins plugin."
    }

    /**
     * The [GameJobProperty] can only be added to jobs with the [jobType] [FreeStyleProject], [WorkflowJob] and
     * [AbstractMavenProject]. For other [jobType]s have a look at [GameMultiBranchProperty] and
     * [GameOrganizationFolderProperty].
     *
     * @see JobPropertyDescriptor.isApplicable
     */
    override fun isApplicable(jobType: Class<out Job<*, *>?>): Boolean {
        return jobType == FreeStyleProject::class.java || jobType == WorkflowJob::class.java
                || AbstractMavenProject::class.java.isAssignableFrom(jobType)
    }

    /**
     * Returns a new instance of a [GameJobProperty] during creation and saving of a job.
     *
     * @see JobPropertyDescriptor.newInstance
     */
    override fun newInstance(req: StaplerRequest?, formData: JSONObject): JobProperty<*>? {
        return if (req == null || req.findAncestor(AbstractItem::class.java).getObject() == null) null
        else GameJobProperty(
            req.findAncestor(AbstractItem::class.java).getObject() as AbstractItem,
            formData.getBoolean(Constants.FormKeys.ACTIVATED),
            formData.getBoolean(Constants.FormKeys.SHOW_LEADERBOARD),
            formData.getBoolean(Constants.FormKeys.SHOW_STATISTICS),
            if (formData.getValue(Constants.FormKeys.CHALLENGES_COUNT) is Int)
                formData.getInt(Constants.FormKeys.CHALLENGES_COUNT) else Constants.Default.CURRENT_CHALLENGES,
            if (formData.getValue(Constants.FormKeys.QUEST_COUNT) is Int)
                formData.getInt(Constants.FormKeys.QUEST_COUNT) else Constants.Default.CURRENT_QUESTS,
            if (formData.getValue(Constants.FormKeys.STORED_CHALLENGES_COUNT) is Int)
                formData.getInt(Constants.FormKeys.STORED_CHALLENGES_COUNT) else Constants.Default.STORED_CHALLENGES,
            formData.getBoolean(Constants.FormKeys.CAN_SEND_CHALLENGE),
            if (formData.getValue(Constants.FormKeys.SEARCH_COMMIT_COUNT) is Int)
                formData.getInt(Constants.FormKeys.SEARCH_COMMIT_COUNT) else Constants.Default.SEARCH_COMMIT_COUNT,
            formData.getString(Constants.FormKeys.PIT_CONFIGURATION),
            formData.getBoolean(Constants.FormKeys.SHOW_PIT_OUTPUT)
        )
    }
}
